lisa/src/main/java/it/unive/lisa/analysis/heap/PointBasedHeap.java

```diff
@@ -0,0 +1,247 @@
+					result = result.lub(
+							mk(sem2, new BinaryExpression(expression.getTypes(), expr1, expr2, binary.getOperator())));
+			return result;
+		}
+
+		if (expression instanceof ValueIdentifier)
+			if (allocationSites.containsKey(expression))
+				return new PointBasedHeap(allocationSites.get(expression), allocationSites);
+
+		if (expression instanceof ValueExpression)
+			return mk(this, (ValueExpression) expression);
+
+		return top();
+	}
+
+	@Override
+	public PointBasedHeap assign(Identifier id, SymbolicExpression expression, ProgramPoint pp)
+			throws SemanticException {
+
+		if (expression instanceof HeapIdentifier) {
+			HashMap<Identifier, Collection<ValueExpression>> sites = new HashMap<>(allocationSites);
+			HashSet<ValueExpression> v = new HashSet<>();
+			v.add((ValueExpression) expression);
+			sites.put(id, v);
+			return new PointBasedHeap(Collections.singleton((ValueExpression) expression), sites);
+		}
+
+		return smallStepSemantics(expression, pp);
+	}
+
+	@Override
+	public PointBasedHeap assume(SymbolicExpression expression, ProgramPoint pp) throws SemanticException {
+		// we just rewrite the expression if needed
+		return smallStepSemantics(expression, pp);
+	}
+
+	@Override
+	public PointBasedHeap forgetIdentifier(Identifier id) throws SemanticException {
+		Map<Identifier, Collection<
+				ValueExpression>> sites = new HashMap<Identifier, Collection<ValueExpression>>(allocationSites);
+		sites.remove(id);
+		return new PointBasedHeap(rewritten, sites);
+	}
+
+	@Override
+	public Satisfiability satisfies(SymbolicExpression expression, ProgramPoint pp) throws SemanticException {
+		// we leave the decision to the value domain
+		return Satisfiability.UNKNOWN;
+	}
+
+	@Override
+	public String representation() {
+		HashSet<ValueExpression> res = new HashSet<ValueExpression>();
+		for (Collection<ValueExpression> s : allocationSites.values())
+			res.addAll(s);
+		return res.toString();
+	}
+
+	@Override
+	public PointBasedHeap top() {
+		return TOP;
+	}
+
+	@Override
+	public PointBasedHeap bottom() {
+		return BOTTOM;
+	}
+
+	@Override
+	public Collection<ValueExpression> getRewrittenExpressions() {
+		return rewritten;
+	}
+
+	@Override
+	public List<HeapReplacement> getSubstitution() {
+		return Collections.emptyList();
+	}
+
+	private PointBasedHeap mk(PointBasedHeap reference, ValueExpression expression) {
+		return new PointBasedHeap(Collections.singleton(expression), reference.allocationSites);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	protected PointBasedHeap lubAux(PointBasedHeap other) throws SemanticException {
+		Collection<ValueExpression> rewritten = (CollectionUtils.union(this.rewritten, other.rewritten));
+		Map<Identifier, Collection<
+				ValueExpression>> sites = new HashMap<Identifier, Collection<ValueExpression>>(allocationSites);
+
+		for (Map.Entry<Identifier, Collection<ValueExpression>> e : other.allocationSites.entrySet())
```

Comment:
Given the usage of `allocationSites`, using an instance of `FunctionalLattice<Identifier, SetLattice<ValueExpression>>` would simplify the implementation of all lattice operations, while also ensuring consistency.